node/silkworm/common/log.cpp

```diff
@@ -14,45 +14,117 @@
    limitations under the License.
 */
 
+#include <fstream>
+#include <iostream>
+#include <ostream>
+#include <regex>
 #include <thread>
 
 #include <absl/time/clock.h>
 
 #include <silkworm/common/log.hpp>
 
-namespace silkworm {
+#if defined(_WIN32)
+#include <windows.h>
+#if !defined(ENABLE_VIRTUAL_TERMINAL_PROCESSING)
+#define ENABLE_VIRTUAL_TERMINAL_PROCESSING 0x0004
+#endif
+#endif
 
-static teestream log_streams_{std::cerr, null_stream()};
+namespace silkworm::log {
 
-LogLevel log_verbosity_{LogLevel::Info};
-bool log_thread_enabled_{false};
+static Settings settings_{};
 
-static constexpr char const kLogTags_[7][6] = {
-    "TRACE", "DEBUG", "INFO ", "WARN ", "ERROR", "CRIT ", "NONE ",
-};
+static std::unique_ptr<std::fstream> file_{nullptr};
 
-// Log to one or two output streams - typically the console and optional log file.
-void log_set_streams_(std::ostream& o1, std::ostream& o2) { log_streams_.set_streams(o1.rdbuf(), o2.rdbuf()); }
+void init(Settings& settings) {
+    settings_ = settings;
+    if (!settings_.log_file.empty()) {
+        tee_file(std::filesystem::path(settings.log_file));
+    }
 
-std::mutex log_::log_mtx_;
+#if defined(_WIN32)
+    // Change code page to UTF-8 so log characters are displayed correctly in console
+    // and also support virtual terminal processing for coloring output
+    SetConsoleOutputCP(CP_UTF8);
+    HANDLE output_handle = GetStdHandle(STD_OUTPUT_HANDLE);
+    if (output_handle != INVALID_HANDLE_VALUE) {
+        DWORD mode = 0;
+        if (GetConsoleMode(output_handle, &mode)) {
+            mode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
+            SetConsoleMode(output_handle, mode);
+        }
+    }
+#endif
+}
 
-std::ostream& log_::header_(LogLevel level) {
-    log_streams_ << kLogTags_[static_cast<int>(level)] << "["
-                 << absl::FormatTime("%m-%d|%H:%M:%E3S", absl::Now(), absl::LocalTimeZone()) << "]";
-    if (log_thread_enabled_) {
-        log_streams_ << " " << std::this_thread::get_id();
+void tee_file(std::filesystem::path path) {
```

Comment:
Currently this method is not thread-safe. I guess it's fine, but it should be documented.

Also, it's better pass `path` by `const &`.

Reply:
About thread safety see same comment for verbosity.

Reply:
Done in 8bb730b